client/order/epochqueue.go

```diff
@@ -0,0 +1,190 @@
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"sync"
+
+	"decred.org/dcrdex/dex/msgjson"
+	"decred.org/dcrdex/dex/order"
+	"github.com/decred/dcrd/crypto/blake256"
+)
+
+// EpochQueue represents a client epoch queue.
+type EpochQueue struct {
+	orders             map[order.OrderID]*msgjson.EpochOrderNote
+	ordersMtx          sync.Mutex
+	seedIndex          []order.OrderID
+	seedIndexMtx       sync.Mutex
+	commitmentIndex    []order.OrderID
+	commitmentIndexMtx sync.Mutex
+}
+
+// NewEpochQueue creates a client epoch queue.
+func NewEpochQueue() *EpochQueue {
+	return &EpochQueue{
+		orders:          make(map[order.OrderID]*msgjson.EpochOrderNote),
+		seedIndex:       make([]order.OrderID, 0),
+		commitmentIndex: make([]order.OrderID, 0),
+	}
+}
+
+// Reset clears the epoch queue. This should be called when a new epoch begins.
+func (eq *EpochQueue) Reset() {
+	eq.ordersMtx.Lock()
+	eq.orders = make(map[order.OrderID]*msgjson.EpochOrderNote)
+	eq.ordersMtx.Unlock()
+
+	eq.seedIndexMtx.Lock()
+	eq.seedIndex = make([]order.OrderID, 0)
+	eq.seedIndexMtx.Unlock()
+
+	eq.commitmentIndexMtx.Lock()
+	eq.commitmentIndex = make([]order.OrderID, 0)
+	eq.commitmentIndexMtx.Unlock()
+}
+
+// Queue appends the provided note to the epoch queue.
+func (eq *EpochQueue) Queue(note *msgjson.EpochOrderNote) {
+	var oid order.OrderID
+	copy(oid[:], note.OrderID[:])
+
+	eq.ordersMtx.Lock()
+	defer eq.ordersMtx.Unlock()
+
+	eq.orders[oid] = note
+
+	// Update the seed sort order per order ids.
+	eq.seedIndexMtx.Lock()
+	si := sort.Search(len(eq.seedIndex), func(i int) bool {
+		order := eq.orders[eq.seedIndex[i]]
+		return bytes.Compare(order.OrderID[:], oid[:]) > 0
+	})
+	eq.seedIndex = append(eq.seedIndex, order.OrderID{})
+	copy(eq.seedIndex[si+1:], eq.seedIndex[si:])
+	eq.seedIndex[si] = oid
+	eq.seedIndexMtx.Unlock()
+
+	// Update the commitment sort order per order commitments.
+	eq.commitmentIndexMtx.Lock()
+	ci := sort.Search(len(eq.commitmentIndex), func(i int) bool {
+		order := eq.orders[eq.commitmentIndex[i]]
+		return bytes.Compare(order.Commitment[:], note.Commitment[:]) > 0
+	})
+	eq.commitmentIndex = append(eq.commitmentIndex, order.OrderID{})
+	copy(eq.commitmentIndex[ci+1:], eq.commitmentIndex[ci:])
+	eq.commitmentIndex[ci] = oid
+	eq.commitmentIndexMtx.Unlock()
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Size() int {
+	eq.ordersMtx.Lock()
+	size := len(eq.orders)
+	eq.ordersMtx.Unlock()
+	return size
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Exists(oid order.OrderID) bool {
+	eq.ordersMtx.Lock()
+	_, ok := eq.orders[oid]
+	eq.ordersMtx.Unlock()
+	return ok
+}
+
+// GenerateMatchProof calculates the sorting seed used in order matching as well as the
+// commitment checksum from the provided epoch queue preimages and misses.
+func (eq *EpochQueue) GenerateMatchProof(preimages []order.Preimage, misses []string) (int64, msgjson.Bytes, error) {
```

Comment:
Why a string instead of order.OrderID for misses?

Reply:
The spec indicated a string slice. Will update to order.OrderID slice.

Reply:
I suppose it depends on who is calling `GenerateMatchProof`, but I don't think it needs to know about the msgjson type where the spec comes in to play.  The json serializations for byte slices/arrays are hexadecimal-encoded strings, but perhaps the spec could use updates in spots to make this explicit everywhere.

Reply:
@dnldd the spec only describes protocols for communication. There is no byte arrays in JSON, so they are expressed as hex strings. But they are order IDs, and we should be reflecting that in our implementation.